src/main/java/org/manifold/compiler/front/MappedArray.java

```diff
@@ -0,0 +1,163 @@
+package org.manifold.compiler.front;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+
+import java.util.*;
+
+public class MappedArray<K, V> implements Iterable<MappedArray<K, V>.Entry> {
+
+  private List<V> data;
+  private BiMap<K, Integer> keyedIndices = HashBiMap.create();
+
+  public MappedArray(Map<K, V> data) {
+    this.data = new ArrayList<>(data.size());
+
+    int i = 0;
+    for (Map.Entry<K, V> entry : data.entrySet()) {
+      this.data.add(entry.getValue());
+      keyedIndices.put(entry.getKey(), i);
+    }
+  }
+
+  public MappedArray(List<V> data) {
+    this.data = data;
```

Comment:
The way I would typically use this kind of constructor (as long as it's documented, which it should be) is to consume a different data structure.  That saves time if the user is never going to use the data structure they passed into the constructor again.  And it leaves them free to do the shallow or deep copy if it's appropriate.  Since we never use the lists we pass into the constructor again I didn't see any reason to make a copy.  But if you think a deep copy is the most useful implementation I'm totally fine with switching to that.